tracee-ebpf/tracee/tracee.go

```diff
@@ -1108,6 +1155,40 @@ func (t *Tracee) processEvent(ctx *context, args map[argTag]interface{}) error {
 	return nil
 }
 
+func getFileHash(fileName string) string {
+	f, _ := os.Open(fileName)
+	if f != nil {
+		defer f.Close()
+		h := sha256.New()
+		_, _ = io.Copy(h, f)
+		return hex.EncodeToString(h.Sum(nil))
+	}
+	return ""
+}
+
+func (t *Tracee) updateProfile(sourceFilePath string, executionTs uint64) {
+	if pf, ok := t.profiledFiles[sourceFilePath]; !ok {
+		t.profiledFiles[sourceFilePath] = profilerInfo{
```

Comment:
Possible race condition when updating this map

Reply:
I tried to think about the various ways today and there are some that I can think of (for same key):

1. Write after Read hazard: This map could be written before reading. I believe this is acceptable.
2. Read after Write hazard. This map is read prior to being written. In this case we would under count the items. This is an issue.
3. Write after Write hazard: This is fine since the counter is only incrementing up.

But regardless, your suggestion made me use the race detector and I found two unrelated races elsewhere in the code which I filed an issue for here https://github.com/aquasecurity/tracee/issues/756

Reply:
Yes, I was referring to case number 2.
I think that a simple mutex for writing will solve this problem, so it is only needed in this function.

Reply:
I thought about this a bit more. Since the read is done once per invocation of this function I don't think there's a race condition here as we only increment the Times counter. 

Here's a simple race test I wrote to demonstrate it:
```
// +build race !unit

package tracee

import (
	"fmt"
	"io/ioutil"
	"os"
	"path/filepath"
	"sync"
	"testing"

	"github.com/stretchr/testify/require"
)

func Test_updateProfile_race(t *testing.T) {
	trc := Tracee{
		profiledFiles: make(map[string]profilerInfo),
	}

	d, err := ioutil.TempDir("", "Test_updateProfile_dir-*")
	require.NoError(t, err)

	f, err := ioutil.TempFile(d, "Test_updateProfile-*")
	require.NoError(t, err)
	defer os.RemoveAll(f.Name())

	captureFileID := fmt.Sprintf("%s.%s", d, filepath.Base(f.Name()))

	//run many concurrent writes
	var wgW sync.WaitGroup
	wgW.Add(10)
	go func() {
		for i := 0; i < 10; i++ {
			trc.updateProfile(captureFileID, 123)
			wgW.Done()
		}
	}()
	wgW.Wait()

	require.Equal(t, profilerInfo{
		Times:            10,
		FirstExecutionTs: 123,
	}, trc.profiledFiles[captureFileID])
}
```

And the output:
```
GOOS=linux GOARCH=amd64 CC=clang CGO_CFLAGS="-I /usr/vagrant/repos/tracee/tracee-ebpf/dist/libbpf/usr/include" CGO_LDFLAGS="/usr/vagrant/repos/tracee/tracee-ebpf/dist/libbpf/libbpf.a" go test -tags=race -race -v ./tracee
=== RUN   Test_updateProfile_race
--- PASS: Test_updateProfile_race (0.00s)
PASS
ok      github.com/aquasecurity/tracee/tracee-ebpf/tracee       0.021s

```

Reply:
I think that 10 counts is not enough to see the race condition.
Here is the case I was thinking about:

Thread A:
1. calls updateProfile
2. pf, ok := t.profiledFiles[sourceFilePath]

Thread B:
1. calls updateProfile
2. pf, ok := t.profiledFiles[sourceFilePath]
3. pf.Times = pf.Times + 1 
4. t.profiledFiles[sourceFilePath] = pf

Thread A:
1. pf.Times = pf.Times + 1 
2. t.profiledFiles[sourceFilePath] = pf

After this sequence, pf.Times will be under count by 1.

I don't think that this issue is that important, as there is nothing functional here (only profiling), but if it is something that can be easilly fixed - why not?

Reply:
@yanivagman what are thread A and thread B? updateProfile is called from ProcessEvent which is a single thread. Can you please elaborate?

Reply:
You are right. My mistake.